File: app/src/main/java/com/pleon/buyt/viewmodel/AddItemViewModel.java
package com.pleon.buyt.viewmodel;

import android.app.Application;

import com.pleon.buyt.database.repository.AddItemRepository;
import com.pleon.buyt.model.Category;
import com.pleon.buyt.model.Item;
import com.pleon.buyt.model.Store;

import java.util.Date;
import java.util.List;

import androidx.lifecycle.AndroidViewModel;

public class AddItemViewModel extends AndroidViewModel {

    private AddItemRepository repository;
    private Category itemCategory = Category.GROCERY;
    private List<Store> storeList;
    private Store store;
    private Date purchaseDate = new Date();
    private int itemOrder;

    public AddItemViewModel(Application application) {
        super(application);
        repository = new AddItemRepository(application);
    }

    public void addItem(Item item) {
        repository.addItem(item);
    }

    public void addPurchasedItem(Item item, Store store, Date purchaseDate) {
        repository.addPurchasedItem(item, store, purchaseDate);
    }

    public Category getCategory() {
        return itemCategory;
    }

    public void setItemCategory(Category itemCategory) {
        this.itemCategory = itemCategory;
    }

    public List<Store> getStoreList() {
        return storeList;
    }

    public void setStoreList(List<Store> storeList) {
        this.storeList = storeList;
    }

    public Store getStore() {
        return store;
    }

    public void setStore(Store store) {
        this.store = store;
    }

    public Date getPurchaseDate() {
        return purchaseDate;
    }

    public void setPurchaseDate(Date purchaseDate) {
        this.purchaseDate = purchaseDate;
    }

    public int getItemOrder() {
        return itemOrder;
    }

    public void setItemOrder(int itemOrder) {
        this.itemOrder = itemOrder;
    }
}

File: app/src/main/kotlin/com/pleon/buyt/viewmodel/AddItemViewModel.kt
package com.pleon.buyt.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.pleon.buyt.R
import com.pleon.buyt.database.dto.ItemNameCat
import com.pleon.buyt.model.Category
import com.pleon.buyt.model.Category.GROCERY
import com.pleon.buyt.model.Item
import com.pleon.buyt.model.Store
import com.pleon.buyt.repository.AddItemRepository
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.launch
import java.util.*

class AddItemViewModel(app: Application, val repository: AddItemRepository) : AndroidViewModel(app) {

    init {
        viewModelScope.launch {
            storeList = repository.getAllStores()
        }
    }

    val defaultQuantity = 1L
    val defaultInitialItemCategory = GROCERY
    lateinit var storeList: List<Store>
    var category = GROCERY
    var purchaseDate = Date()
    var store: Store? = null

    private val defaultNameCats = flow {
        fun String.toNameCat() =
            ItemNameCat(substringBefore(":"), Category.valueOf(substringAfter(":")))
        app.resources.openRawResource(R.raw.item_names)
            .reader()
            .useLines { lines ->
                emit(lines.map { it.toNameCat() }.toList())
            }
    }.flowOn(Dispatchers.IO)

    /**
     * Could also have written it like this:
     * ```
     * val x = flow1.combine(flow2) { f1, f2 -> ... }
     * ```
     */
    val itemNameCats = combine(defaultNameCats, repository.itemNameCats) { default, database ->
        (default + database).associateBy({ it.name }, { it.category })
    }.flowOn(Dispatchers.Default)

    fun addItem(item: Item, isBought: Boolean) = viewModelScope.launch {
        if (isBought) {
            repository.addPurchasedItem(item, store!!, purchaseDate)
        } else {
            repository.addItem(item)
        }
    }

    fun resetValues() {
        purchaseDate = Date()
        category = defaultInitialItemCategory
        store = null
    }
}
